Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,746 @@
+        |6, 7|
+    |
+];
+Message($"{data[1, 1, 0]}"); // ← 6
+```
+
+Example 2:
+Mixing 1D and multidimensional arrays.
+
+```qsharp
+// Here, we declare the first two levels of indexing as being rectangular,
+// resulting in a three-dimensional array.
+let data = [
+    |
+        // After rows, [] denote arrays as elements again.
+        | [0], [1, 2] |,
+        | [3, 4, 5], [6, 7, 8, 9] |
+    |,
+
+    |
+        | [10, 11, 12, 13, 14], [15, 16, 17, 18, 19, 20] |,
+        | [21, 22, 23, 24, 25, 26, 27], [28, 29, 30, 31, 32, 33, 34, 35] |
+    |
+];
+// data is of type [||[Int]||] (that is, a three-dimensional array of arrays of integers)
+// In the zero-th "plane", first "row", zero-th "column", get the element with index 2.
+Message($"{data[0, 1, 0][2]}"); // ← 5
+```
+
+Example 3:
+Using expressions as subarrays of multidimensional arrays results in a compile-time error.
+
+```qsharp
+let a = [2, 3];
+// The following is a compile-time error, as `|0, 1|` is not an expression
+// independent of its surrounding `[]` brackets.
+let data = [
+    | 0, 1 |,
+    a
+];
+// Using a new library function that can `fail` at runtime works, however.
+let data = Concatenated2(0, // concatenate along the 0th (row) axis
+    [ | 0, 1 | ],
+    a
+);
+// data: [|Int|] = [
+//     |0, 1|,
+//     |2, 3|
+// ];
+```
+
+Example 4:
+Using the copy-and-update operator to manipulate multidimensional arrays.
+
+```qsharp
+function ElementaryMatrix(
+    (nRows : Int, nCols : Int), (idxRow : Int, idxCol : Int)
+) : [|Double|] {
+    return [| 0.0, size = (nRows, nCols) |] w/ (idxRow, idxCol) <- 1.0;
+}
+```
+
+Example 5:
+Slicing multidimensional arrays.
+
+```qsharp
+let data = [
+    |0, 1, 2|,
+    |3, 4, 5|,
+    |6, 7, 8|
+];
+
+// Slicing does not reduce the dimensionality of the resulting array.
+let corners = data[0..2..2, 0..2..2];
+// corners = [|0, 2|, |6, 8|] and is of type [|Int|]
+
+// Subscripting by an Int reduces the dimensionality of the resulting array;
+// here, since our index has one Int, the dimensionality reduces from
+// [|Int|] to [Int].
+let firstRow = data[0, ...];
+// firstRow = [0, 1, 2]
+// The same pattern holds no matter which index we subscript with an Int.
+let firstColumn = data[..., 0];
+// firstColumn = [0, 3, 6]
+
+let data3 = [
+    |
+        |0, 1, 2|,
+        |3, 4, 5|,
+        |6, 7, 8|
+    |,
+
+    |
+        |9, 10, 11|,
+        |12, 13, 14|,
+        |15, 16, 17|
+    |,
+
+    |
+        |18, 19, 20|,
+        |21, 22, 23|,
+        |24, 25, 26|
+    |
+];
+let corners3 = data3[0..2..2, 0..2..2, 0..2..2];
+// corners3 is of type [||Int||]
+let firstPlane = data3[0, ..., ...];
+// firstPlane is of type [|Int|]
+let firstRowOfFirstPlane = data3[0, 0, ...];
+// firstRowOfFirstPlane = [0, 1, 2] and is of type [Int]
+```
+
+Example 6:
+Using multidimensional slices in copy-and-update expressions.
+
+```qsharp
+let zeros =[|0, size = (3, 3)|];
+
+let withCorners = zeros w/ (0..2..2, 0..2..2) <- [|1, 2|, |3, 4|];
```

Comment:
One thing that is not explicitly stated and I think fine to just add here as a comment, is that the right hand side `[|1,2|,|3,4|]` must be a multidimensional array opposed to an array of arrays.